### PR TITLE
feat: add utilities to detect gcp

### PR DESCRIPTION
**Pull request by ddelgrosso1**
Add GCP detection utilities.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655)

### CODE REVIEW

**Review by scotthart — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655)* status: 0 of 11 files reviewed, 8 unresolved discussions (waiting on @bajajneha27, @ddelgrosso1, and @frankyn)

___
*[`google/cloud/internal/detect_gcp.h` line 43 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qm38CE0omuuJ5OkgW:-O4qkp208d7kmR7qhlJ9:b-98iei8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp.h#L43)):*
> ```objc
>     HKEY key = HKEY_LOCAL_MACHINE;
>     std::string sub_key = "SYSTEM\\HardwareConfig\\Current";
>     std::string value_key = "SystemProductName";
> ```

I would expect that these types would be defined as part of the derived class that implements this interface, not located in the base class.
___
*[`google/cloud/internal/detect_gcp.h` line 59 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qmES1CwpXDvxIJVT3:-O4qmES1CwpXDvxIJVT4:bcrx43d) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp.h#L59)):*
> ```objc
> 
>  private:
>   virtual std::string GetBiosInformation() = 0;
> ```

It's unusual for an interface class to have virtual private functions. If the derived class needs to implement such a function, it can. It shouldn't be required to.
___
*[`google/cloud/internal/detect_gcp.h` line 61 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qma-N-nEzz_VXsP6I:-O4qma-O9358MIqOfbQC:b-57hw5y) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp.h#L61)):*
> ```objc
>   virtual std::string GetBiosInformation() = 0;
> };
> ```

Consider adding a factory function that creates the appropriate GcpDetectorImpl and returns it as a GcpDetector.
___
*[`google/cloud/internal/detect_gcp_impl.h` line 29 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qn3Hn5xVODhQiHGNO:-O4qn3Ho0J6bM8jkPV3s:blyerca) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp_impl.h#L29)):*
> ```objc
> class GcpDetectorImpl : GcpDetector {
>  public:
>   explicit GcpDetectorImpl(GcpDetectorConfig const& config)
> ```

For a constructor, take config by value, not const&.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 90 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qpPQm49jLDhI0dtEF:-O4qpPQm49jLDhI0dtEG:b-izgjm9) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp_impl.cc#L90)):*
> ```cpp
>                        // Use _dupenv_s here instead of getenv.
>                        // MSVC throws a security error on getenv.
>                        auto result = _dupenv_s(&buf, &size, env_var.c_str());
> ```

Use the functions in google/cloud/internal/getenv.h for interacting with environment variables.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 101 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qpZcP29mll1qzTf43:-O4qpZcQEjKxpJKnpzuz:b-izgjm9) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp_impl.cc#L101)):*
> ```cpp
>                      this->config_.env_variables.end(),
>                      [](std::string const& env_var) {
>                        return std::getenv(env_var.c_str()) != nullptr;
> ```

Use the functions in google/cloud/internal/getenv.h for interacting with environment variables.
___
*[`google/cloud/internal/detect_gcp_test.cc` line 47 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qpjC149jYwLyAUwNk:-O4qpjC149jYwLyAUwNl:b-fp0aa) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp_test.cc#L47)):*
> ```cpp
> 
> #ifdef _WIN32
> std::string const& parent_key = "SOFTWARE\\GoogleCloudCpp";
> ```

auto constexpr parent_key = "..." should save you from having to call c_str() a lot.
___
*[`google/cloud/internal/detect_gcp_test.cc` line 180 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qqPLj9WN8RPadkzuP:-O4qqPLj9WN8RPadkzuQ:b-mwgthh) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp_test.cc#L180)):*
> ```cpp
> 
> #ifdef _WIN32
>   _putenv_s(env_vars[0].c_str(), "VALUE");
> ```

Use `ScopedEnvironment` from google/cloud/testing_util/scoped_environment.h for testing with env vars.

**Review by scotthart — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655)* status: 0 of 11 files reviewed, 14 unresolved discussions (waiting on @bajajneha27, @ddelgrosso1, and @frankyn)

___
*[`google/cloud/internal/detect_gcp_impl.h` line 22 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4zxwUm0v7WMbIdjFLt:-O4zxwUm0v7WMbIdjFLu:b-wssqgw) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63dad2d12bcd65171b1374cac7be379cbb62f1f3/google/cloud/internal/detect_gcp_impl.h#L22)):*
> ```objc
> #include <memory>
> #include <string>
> #include <utility>
> ```

You may be able to remove this header.
___
*[`google/cloud/internal/detect_gcp_impl.h` line 58 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4zxj9R2HP_8GaeZwm8:-O4zxj9R2HP_8GaeZwm9:bp9ep9n) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63dad2d12bcd65171b1374cac7be379cbb62f1f3/google/cloud/internal/detect_gcp_impl.h#L58)):*
> ```objc
> };
> 
> std::shared_ptr<GcpDetector> MakeGcpDetector();
> ```

I would move this function declaration to detect_gcp.h.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 24 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4zz1caBHaTdj2L0l5-:-O4zz1caBHaTdj2L0l50:b-orvl3a) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63dad2d12bcd65171b1374cac7be379cbb62f1f3/google/cloud/internal/detect_gcp_impl.cc#L24)):*
> ```cpp
> #include <iostream>
> #include <memory>
> #include <regex>
> ```

I think this is unused.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 40 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4zy41b6RD5xWP5dNZs:-O4zy41b6RD5xWP5dNZt:b-6kmh5a) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63dad2d12bcd65171b1374cac7be379cbb62f1f3/google/cloud/internal/detect_gcp_impl.cc#L40)):*
> ```cpp
> #ifdef _WIN32
>   DWORD size{};
>   LONG result = RegGetValueA(this->config_.key, this->config_.sub_key.c_str(),
> ```

As this is a member function, you don't need to specify the `this->`, you can just refer to the member variable directly: `config_.key`
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 44 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4zzX6u2wBzxuOT2hbb:-O4zzX6u2wBzxuOT2hbc:b-isqz1t) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63dad2d12bcd65171b1374cac7be379cbb62f1f3/google/cloud/internal/detect_gcp_impl.cc#L44)):*
> ```cpp
>                              nullptr, nullptr, &size);
> 
>   if (result != ERROR_SUCCESS) return "";
> ```

So here and other places, we're returning an empty string when an error is encountered. It's probably useful to know which error we're getting, I would recommend returning a `google::cloud::StatusOr<std::string>`. There may arise situations where we expect it to be a gcp machine, but we're saying it's not because the bios information isn't in the correct location. While this may not be a hard error, we could call `GCP_LOG` with the error so that there's some record of what happened.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 61 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O5-0lx_-9Az5QgB8vde:-O5-0lx_-9Az5QgB8vdf:b-xvwodi) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63dad2d12bcd65171b1374cac7be379cbb62f1f3/google/cloud/internal/detect_gcp_impl.cc#L61)):*
> ```cpp
> #else  // _WIN32
>   auto product_name_status =
>       google::cloud::internal::status(this->config_.path);
> ```

We're already in the `google::cloud::internal` namespace, so this (and elsewhere) can be written with just `status(config_.path)`.

**Review by scotthart — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655)* status: 0 of 11 files reviewed, 16 unresolved discussions (waiting on @bajajneha27, @ddelgrosso1, and @frankyn)

___
*[`google/cloud/internal/detect_gcp_impl.cc` line 46 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O5Euk6L9BooZoDgD-kG:-O5Euk6L9BooZoDgD-kH:b6od0jt) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/74bad67acb9ba7536df3c6c87306d66693eac5cf/google/cloud/internal/detect_gcp_impl.cc#L46)):*
> ```cpp
>   if (result != ERROR_SUCCESS)
>     return StatusOr<std::string>(
>         Status(StatusCode::kUnknown, "error querying registry value"));
> ```

We have some helper functions in `google/cloud/internal/make_status.h` for creating errors. So something like:
```
return internal::MakeInternalError("error querying registry value", GCP_ERROR_INFO().WithMetadata("key", config_.key).WithMetadata("sub_key", config_.sub_key));
```
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 84 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O5Ew1zA7aaKFQIKhFrM:-O5Ew1zA7aaKFQIKhFrN:b-pkg5yk) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/74bad67acb9ba7536df3c6c87306d66693eac5cf/google/cloud/internal/detect_gcp_impl.cc#L84)):*
> ```cpp
> bool GcpDetectorImpl::IsGoogleCloudBios() {
>   auto status = GetBiosInformation();
>   if (!status.ok()) return false;
> ```

I would add a log call for any errors so the user/tester can then enable logging and see if something unexpected is happening.

```
if (!status) {
  GCP_LOG(WARNING) << std::move(status).status();
  return false;
}
```

**Review by ddelgrosso1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655)* status: 0 of 11 files reviewed, 16 unresolved discussions (waiting on @bajajneha27, @frankyn, and @scotthart)

___
*[`google/cloud/internal/detect_gcp.h` line 43 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qm38CE0omuuJ5OkgW:-O5FB__02MXqJRXq7S6X:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp.h#L43)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

I would expect that these types would be defined as part of the derived class that implements this interface, not located in the base class.
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp.h` line 59 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qmES1CwpXDvxIJVT3:-O5FBa436O5x6TzCnLao:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp.h#L59)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

It's unusual for an interface class to have virtual private functions. If the derived class needs to implement such a function, it can. It shouldn't be required to.
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp.h` line 61 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qma-N-nEzz_VXsP6I:-O5FBaRU8fcoIuVZ09aK:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp.h#L61)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Consider adding a factory function that creates the appropriate GcpDetectorImpl and returns it as a GcpDetector.
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_impl.h` line 29 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qn3Hn5xVODhQiHGNO:-O5FBbYcApK8Gziy_iEF:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp_impl.h#L29)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

For a constructor, take config by value, not const&.
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_impl.h` line 22 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4zxwUm0v7WMbIdjFLt:-O5FBaxZ8GqUzqBX9YdH:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63dad2d12bcd65171b1374cac7be379cbb62f1f3/google/cloud/internal/detect_gcp_impl.h#L22)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

You may be able to remove this header.
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_impl.h` line 58 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4zxj9R2HP_8GaeZwm8:-O5FBcAR9En5lS713ekE:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63dad2d12bcd65171b1374cac7be379cbb62f1f3/google/cloud/internal/detect_gcp_impl.h#L58)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

I would move this function declaration to detect_gcp.h.
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 90 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qpPQm49jLDhI0dtEF:-O5FBjZ74DvVfpjhB-AD:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp_impl.cc#L90)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Use the functions in google/cloud/internal/getenv.h for interacting with environment variables.
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 101 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qpZcP29mll1qzTf43:-O5FBjxN7XAEAYbvIi0w:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp_impl.cc#L101)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Use the functions in google/cloud/internal/getenv.h for interacting with environment variables.
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 24 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4zz1caBHaTdj2L0l5-:-O5FBct06m9tutH_iD3P:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63dad2d12bcd65171b1374cac7be379cbb62f1f3/google/cloud/internal/detect_gcp_impl.cc#L24)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

I think this is unused.
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 40 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4zy41b6RD5xWP5dNZs:-O5FBdOJ4MPeLRgyX2_y:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63dad2d12bcd65171b1374cac7be379cbb62f1f3/google/cloud/internal/detect_gcp_impl.cc#L40)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

As this is a member function, you don't need to specify the `this->`, you can just refer to the member variable directly: `config_.key`
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 44 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4zzX6u2wBzxuOT2hbb:-O5FBeJRB0xe-bRnFK1D:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63dad2d12bcd65171b1374cac7be379cbb62f1f3/google/cloud/internal/detect_gcp_impl.cc#L44)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

So here and other places, we're returning an empty string when an error is encountered. It's probably useful to know which error we're getting, I would recommend returning a `google::cloud::StatusOr<std::string>`. There may arise situations where we expect it to be a gcp machine, but we're saying it's not because the bios information isn't in the correct location. While this may not be a hard error, we could call `GCP_LOG` with the error so that there's some record of what happened.
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 61 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O5-0lx_-9Az5QgB8vde:-O5FBhVLCG_Ftd8beLdk:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/63dad2d12bcd65171b1374cac7be379cbb62f1f3/google/cloud/internal/detect_gcp_impl.cc#L61)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

We're already in the `google::cloud::internal` namespace, so this (and elsewhere) can be written with just `status(config_.path)`.
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 46 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O5Euk6L9BooZoDgD-kG:-O5FBeuy-y0Z8oVUon0Y:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/74bad67acb9ba7536df3c6c87306d66693eac5cf/google/cloud/internal/detect_gcp_impl.cc#L46)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

We have some helper functions in `google/cloud/internal/make_status.h` for creating errors. So something like:
```
return internal::MakeInternalError("error querying registry value", GCP_ERROR_INFO().WithMetadata("key", config_.key).WithMetadata("sub_key", config_.sub_key));
```
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_impl.cc` line 84 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O5Ew1zA7aaKFQIKhFrM:-O5FBj-A6ywL8BDclHLT:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/74bad67acb9ba7536df3c6c87306d66693eac5cf/google/cloud/internal/detect_gcp_impl.cc#L84)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

I would add a log call for any errors so the user/tester can then enable logging and see if something unexpected is happening.

```
if (!status) {
  GCP_LOG(WARNING) << std::move(status).status();
  return false;
}
```
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_test.cc` line 47 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qpjC149jYwLyAUwNk:-O5FBkbK5J1ksaNtFyib:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp_test.cc#L47)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

auto constexpr parent_key = "..." should save you from having to call c_str() a lot.
</blockquote></details>

Done.
___
*[`google/cloud/internal/detect_gcp_test.cc` line 180 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655#-O4qqPLj9WN8RPadkzuP:-O5FBl8QANtoQV_F6k26:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a080618998ac3fdfc83cc6635155034627acecdc/google/cloud/internal/detect_gcp_test.cc#L180)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Use `ScopedEnvironment` from google/cloud/testing_util/scoped_environment.h for testing with env vars.
</blockquote></details>

Done.

File: google/cloud/internal/detect_gcp_impl.cc [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                     });
+}
+
+std::shared_ptr<GcpDetector> MakeGcpDetector() {
+  auto config = GcpDetectorImpl::GcpDetectorConfig{};
+  config.env_variables = {"CLOUD_RUN_JOB", "FUNCTION_NAME", "K_SERVICE"};
+#ifdef _WIN32
+  config.key = HKEY_LOCAL_MACHINE;
+  config.sub_key = "SYSTEM\\HardwareConfig\\Current";
+  config.value_key = "SystemProductName";
```
Comment (frankyn) [inline google/cloud/internal/detect_gcp_impl.cc:120]: Is there a ref doc you can link to on how this works?

**Review by frankyn — APPROVED** [review]
LGTM

**Review by scotthart — APPROVED** [review]
> Reviewed 7 of 11 files at r2, 3 of 4 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14655)* status: all files reviewed, 17 unresolved discussions (waiting on @bajajneha27 and @frankyn)

**Review by frankyn — APPROVED** [review]

**Review by bajajneha27 — APPROVED** [review]

**Pull request merged**